### DISCUSSION TITLE
Digging into 500 .bin file

### DISCUSSION THREAD

**Discussion by GerryDazoo** [GerryDazoo/Slinger#428] [General]
Any Linux experts out there that might want to take a look at this?
The file name s_uImage_mips_gz_378.bin smells awful like a Linux ramdisk/boot image. Would anybody like to unpack it and do a quick mount on the file to see what's there? You never know we might be able to add our own executables to the image.
I'd do it myself but I'm kinda busy. 

-Gerry

**Comment by regmar505** [comment]
Where could we find  s_uImage_mips_gz_378.bin and the SDK you talked about in "Digging into 500" ? Would it make sense to have them available here?

**Reply by GerryDazoo** [reply]
get it here
http://home.dazoo.net:64888/slinger/s_uImage_mips_gz_378.bin

**Comment by GerryDazoo** [comment]
Bueller? Bueller?

**Reply by regmar505** [reply]
Sorry, I realize afterward that slingbox uses MIPS architecture, this is not for me. I hope that somebody else could step in.

**Reply by GerryDazoo** [reply]
I'm just looking for what's on the ramdisk, It doesn't matter if it's MIPS unless you try to run it.

**Reply by regmar505** [reply]
I am going nowhere, I suspect this file encrypted or compressed by an unusual program, I cannot find a way to mount it.

**Reply by DVoorhees1957** [reply]
I'm way over my head here, but there's an open source MIPS firmware extraction/modification project at https://github.com/rampageX/firmware-mod-kit. In the readme, under "Extracting Firmware", it says "Automated firmware extraction typically works with most firmware images that employ uImage/TRX firmware headers and use SquashFS or CramFS file systems."

The name of this BIN file (s_uImage_mips_gz_378.bin) suggests that it is indeed a uImage firware image, but I have no idea if the BIN file uses SquashFS or CramFS. If there's someone with more Linux chops than I have, they could try to just do the firmware extraction bit, to see if there's anything meaningful/readable that results...

**Reply by GerryDazoo** [reply]
Thanks for the ol' college try. I was hoping it might be something easy like renaming the bin to .gz and unzipping it. :-) . Don't spend too much time on this for now unless you really really want to.

**Reply by regmar505** [reply]
Thank you for the open source mentioned above
 `./unsquashfs_all.sh s_uImage_mips_gz_378.bin`

./image_parts:
total 5340
-rw-r--r-- 1 root root      24 Dec 13 08:20 hwid.txt
-rw-r--r-- 1 root root 4412784 Dec 13 08:20 squashfs-3-lzma.img
-rw-r--r-- 1 root root 1048576 Dec 13 08:20 vmlinuz

However, I am unable to uncompress it, neither to mount it:
`sudo unsquashfs -f -d tmp squashfs-3-lzma.img`
FATAL ERROR: Can't find a valid SQUASHFS superblock on squashfs-3-lzma.img

I have attached squashfs-3-lzma.img perhaps somebody else could try. I suspect that I have a wrong uncompress version embedded in unsquashfs.

[squashfs-3-lzma.img.zip](https://github.com/GerryDazoo/Slinger/files/10218819/squashfs-3-lzma.img.zip)

**Reply by DVoorhees1957** [reply]
Well, that's progress (of sorts!). Vmlinuz is just the (compressed) Linux kernel executable, so it seems like you're headed the right direction - trying to access the squashfs image.

A couple of thoughts, just based on a quick Google of "squashfs-3-lzma.img":
1. Maybe try unsquashfs-lzma rather than just unsquashfs
2. Maybe try a little bit of tweaking to the unsquashfs-lzma process (see the 3rd entry on the following thread, https://forum.dd-wrt.com/phpBB2/viewtopic.php?t=77040; I'll extract and copy the key comments here):
```
All tools are found in the firmware modkit:

mkdir tmp
untrx <imagefile> tmp/
cd tmp
dd if=segment1 of=squashfs.img bs=16 skip=49152 conv=notrunc
unsquashfs-lzma squashfs.img

And we get a squashfs-root dir with the FS contents.

How I did that (actually not that high magic Wink)
* Create some random, small squashfs images
* Look at them with hexdump
* Learn that 7368 7471 seems to be the magic header of a squashfs lzma image
* Find single occurence of that pattern in the untrx'd segment1
* Copy&unpack it
```

**Reply by GerryDazoo** [reply]
Yahoo. That's good stuff. Interesting they are using Linux. Do ya'll recall how Linksys was sued for violating the GPL and were forced to open source the info people needed to put their own code on their routers.  DD-WRT was born.

**Reply by DVoorhees1957** [reply]
@regmar505 , I cloned the Firmware Mod Kit git code and tried to run the same command you did,
```
./unsquashfs_all.sh s_uImage_mips_gz_378.bin
```
However, I didn't even get as far as you did - the file extraction simply failed (without producing the image_parts directory that you were able to obtain). The screen shot follows. Any ideas on what I'm doing wrong?

![Unsquash failure](https://user-images.githubusercontent.com/13420879/207470428-132f4538-1eef-42e8-a046-fc5c1a36de40.jpg)

**Reply by regmar505** [reply]
I get these results despite the error below.

me@nuc:/tmp/firmware-mod-kit-master $ `cp deprecated/old-extract.sh .`
me@nuc:/tmp/firmware-mod-kit-master $ `sudo  ./old-extract.sh s_uImage_mips_gz_378.bin tmp`

 Firmware Mod Kit (extract) v0.99, (c)2010-2012 Jeremy Collake, ......
 LINUX system detected. Compatibility ok.
 Testing file system of tmp ...
 Building tools ...
./old-extract.sh: 128: ./configure: not found
/tmp/firmware-mod-kit-master
 Build seems successful.
 Preparing working directory ...
 Removing any previous files ...
 Creating directories ...
 Extracting firmware
! untrx failed, trying splitter3
 Not recognized by splitter3
 Attempting raw linux style firmware package (i.e. TEW-632BRP) ...
**Extracting s_uImage_mips_gz_378.bin to tmp**
 Error: filesystem not extracted properly.
  firmware image format not compatible?

me@nuc:/tmp/firmware-mod-kit-master $ `cd tmp`
me@nuc:/tmp/firmware-mod-kit-master/tmp $ `ls -lR`

drwxr-xr-x 2 root root 100 Dec 13 21:33 image_parts
drwxr-xr-x 2 root root  40 Dec 13 21:33 installed_packages

./image_parts:
total 5340
-rw-r--r-- 1 root root      24 Dec 13 21:33 hwid.txt
-rw-r--r-- 1 root root 4412784 Dec 13 21:33 squashfs-3-lzma.img
-rw-r--r-- 1 root root 1048576 Dec 13 21:33 vmlinuz

installed_packages directory is empty. 
Hope this helps

**Reply by DVoorhees1957** [reply]
Thanks, that helped. For some reason, I had to copy old-extract.sh from the deprecated folder to the main firmware-mod-kit folder; after I did that, I get basically the same output (and results) as you.

**Reply by DVoorhees1957** [reply]
Not sure it means anything, but the contents of the hwid.txt file are just the last 24 bytes of the squashfs-3-lzma.img file, so ignore it for the moment. The combined size of the other 2 files, vmlinuz (1048576 bytes) and squashfs-3-lzma.img (4412784 bytes), is equal to the size of the original s_uImage_mips_gz_378.bin file (5461360 bytes) - and the beginning of the vmlinuz file matches the beginning of the s_uImage_mips_gz_378.bin file. So it seems like the "extract" process simply picked a point to split the s_uImage_mips_gz_378.bin file into two parts, the 1st part becoming the vmlinuz file and the 2nd part becoming the squashfs-3-lzma.img file. What if the extraction process simply choose the wrong split point? That might explain why the unsquashfs process isn't working...

**Reply by GerryDazoo** [reply]
That makes sense.

**Reply by DVoorhees1957** [reply]
I've spent the better part of the day trying to understand the LZMA file structure and the uImage file structure... an LZMA file has a 13-byte header, the 1st byte of which is the Properties byte (which usually has the value 93 (x'5D'), the next 4 bytes are the Dictionary Size, and the last 8 bytes are the Uncompressed Size... still working on pinning down a specific format for a uImage file. I've seen uImage files with 64-byte headers, and some with 128-byte headers.

I've also been trying to use the Binwalk took, which is supposed to be able to "parse" most binary files, find interesting/common file signatures, etc... But for the life of me, I can't get Binwalk to find anything useful to report, either on the original BIN file or on the extracted squashfs IMG file. So I feel sorta like I've been spinning my wheels...

**Reply by GerryDazoo** [reply]
Standby, let me have a look through the packet capture of the firmware download. There may be some metadata in there that tells the existing firmware how to unpack the image.
Here's something.

{"payload":"{\"config\":{\"updateVersion\":\"1.8.402\",\"firmwareComponents\":{\"appfsMain\":{\"critical\":false,\"crc\":123,\"order\":1,\"reboot\":false,\"url\":\"www.abcdf.com\",\"size\":123,\"version\":\"0.5.102\"},\"uImageMain\":{\"critical\":false,\"crc\":516463924,\"order\":0,\"reboot\":false,\"url\":\"http://cbfu-prod.slingbox.com/Intrepid/Intrepid_FW_1_8_392/s_uImage_mips_gz_378.bin\",\"size\":5461360,\"version\":\"1.8.378\"}},\"applications\":{\"sbCore\":{\"critical\":false,\"reboot\":true,\"url\":\"http://cbfu-prod.slingbox.com/Intrepid/Intrepid_FW_1_8_402v2/intrepid_sbcore_f_p_1_8_402v2.tar\",\"version\":\"1.8.402\",\"size\":64235520}}}}","header":{"signatureEncoding":"base64","msgType":"plain","signature":"MIAGCSqGSIb3DQEHAqCAMIACAQExDzANBglghkgBZQMEAgEFADCABgkqhkiG9w0BBwEAADGCAg8wggILAgEBMHkwdDELMAkGA1UEBhMCVVMxEzARBgNVBAgTCkNhbGlmb3JuaWExFDASBgNVBAoTC1NsaW5nIE1lZGlhMRMwEQYDVQQDEwpNRCBET1dOIENBMSUwIwYJKoZIhvcNAQkBFhZtZGFkbWluQHNsaW5nbWVkaWEuY29tAgEDMA0GCWCGSAFlAwQCAQUAoGkwGAYJKoZIhvcNAQkDMQsGCSqGSIb3DQEHATAcBgkqhkiG9w0BCQUxDxcNMTQxMTIwMDgwOTQwWjAvBgkqhkiG9w0BCQQxIgQglWihrwzPfsDK5wDEMiMkMejkD4G9NXEPF0PCjcOcpd0wDQYJKoZIhvcNAQEBBQAEggEAYQuM+IVVcODQU976GWQ2dMjh3zeN9GY6m5iVtGqHlsucChTPsoVJ4LHVxAuZhYPhddL3LRXWtHXtm8lFDNRv7dh9ToK7Aw4SgboYDRYXC/hYwneGapllD5hxEbycamnLewmDp/hAa5dpAOjLJm4c7uqHAr6mitd052dVpOaArltrY0JJBwugZ5OFk7NbByh6DosqwtSdz7dLtT5IK/Kec8lRWnqdMPvWJiNImdekI/HIYCxZRjMfr5DWVsO7lZAcCUB8q0ADqwL34nYJsr/3IK/XLtenF01ASYCcwBL5pN8sRp8sZflLUBNjv8D3oYa5SPK2Q1O6Dr++lEIewZx/VgAAAAAAAA==","configEncoding":"none"}}

Holy smoke; I never noticed this before!
http://cbfu-prod.slingbox.com/Intrepid/Intrepid_FW_1_8_402v2/intrepid_sbcore_f_p_1_8_402v2.tar

**Reply by GerryDazoo** [reply]
Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a----         8/19/2014  10:55 AM         452046 App-Mgr_0.0_all.ipk.smime
-a----         8/19/2014  10:55 AM        7057710 kernel-modules_0.0_all.ipk.smime
-a----         8/19/2014  10:55 AM       12729646 lib_0.0_all.ipk.smime
-a----         8/19/2014  10:55 AM         146014 ME_0.0_all.ipk.smime
-a----         8/19/2014  10:55 AM          59233 MTW-App_0.0_all.ipk.smime
-a----         8/19/2014  10:55 AM          79734 mymediaservice_0.0_all.ipk.smime
-a----         8/19/2014  10:55 AM        1140518 PS-App_0.0_all.ipk.smime
-a----         8/19/2014  10:55 AM          28257 scripts_0.0_all.ipk.smime
-a----         8/19/2014  10:55 AM       16405742 uiapp-common_0.0_all.ipk.smime
-a----         8/19/2014  10:55 AM       17185451 uiapps_0.0_all.ipk.smime
-a----         8/19/2014  10:55 AM        8933766 utils_0.0_all.ipk.smime

**Comment by DVoorhees1957** [comment]
@GerryDazoo is there anything more that's useful to do on this thread? My suspicion is that the intrepid_sbcore_f_p_1_8_402v2.tar file you discovered is the core Slingbox code, and that code in turn knows how to load and interpret the s_uImage_mips_gz_378.bin file.
